### PR TITLE
Build API docs with mkdocstrings

### PR DESCRIPTION
**Pull request by tpdorsey**
Builds docs from docstrings as an API guide. 

This is WIP to highlight what's building from existing docstrings. Only builds docs for elements with valid docstrings.

Preview: https://smokeshow.helpmanual.io/2b2n0w590e055e6m6c1g/api/analyzed_type/

## Change Summary

* adds API tab to navigation
* adds a page under API for each public code file
* builds docs from docstrings

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

**Comment by dmontagu** [comment]
Looks good to me other than some minor issues with the visual display. If you want feedback on that let me know, but I figure maybe that makes more sense to do as part of a separate PR.

Is there anything now that will validate the format of the docstrings? I strongly suspect they will not remain consistent with the changes in this PR if nothing checks them. In particular, I am not in the habit of including text on the same line as the initial `"""`, so I'll probably end up merging something not following that pattern if we don't enforce it with linting and/or some other CI job.

**Comment by tpdorsey** [comment]
Including the summary line on the first """ line is recommended style, but not enforced, nor does it affect testing or rendering. So this should not be a problem @dmontagu 

The linter will pick up formatting errors, and mkdocstrings will error on substantive errors (mismatched or incorrect args, for example).

**Comment by tpdorsey** [comment]
Please review

File: pydantic/analyzed_type.py [hunk 1]
```diff
@@ -107,11 +121,34 @@ def __init__(self, __type: Any, *, config: ConfigDict | None = None, _parent_dep
...
+        """Validate a JSON string or bytes against the model.
+
+        Args:
+            __data (str | bytes): The JSON data to validate against the model.
+            strict (bool | None, optional): Whether to strictly check types. Defaults to None.
+            context (dict[str, Any] | None, optional): Additional context to use during validation. Defaults to None.
+
+        Returns:
+            T: The validated object.
+
```
Comment (Kludex) [inline pydantic/analyzed_type.py:150]:
```suggestion
            T: The validated object.
```

File: pydantic/analyzed_type.py [hunk 2]
```diff
@@ -107,11 +121,34 @@ def __init__(self, __type: Any, *, config: ConfigDict | None = None, _parent_dep
...
+        Validate a Python object against the model.
+
+        Args:
+            __object (Any): The Python object to validate against the model.
+            strict (bool | None, optional): Whether to strictly check types. Defaults to None.
+            context (dict[str, Any] | None, optional): Additional context to use during validation. Defaults to None.
+
+        Returns:
+            T: The validated object.
+
```
Comment (Kludex) [inline pydantic/analyzed_type.py:134]:
```suggestion
            T: The validated object.
```

File: pydantic/analyzed_type.py [hunk 3]
```diff
@@ -128,6 +165,25 @@ def dump_python(
...
+            by_alias (bool, optional): Whether to use alias names for field names. Defaults to False.
+            exclude_unset (bool, optional): Whether to exclude unset fields. Defaults to False.
+            exclude_defaults (bool, optional): Whether to exclude fields with default values. Defaults to False.
+            exclude_none (bool, optional): Whether to exclude fields with None values. Defaults to False.
+            round_trip (bool, optional): Whether to output the serialized data in a way that is compatible with
+                deserialization. Defaults to False.
+            warnings (bool, optional): Whether to display serialization warnings. Defaults to True.
+
+        Returns:
+            Any: The serialized object.
```
Comment (Kludex) [inline pydantic/analyzed_type.py:184]: There's an inconsistency on this PR between the `Returns`: some have the return type, and some doesn't.

File: pydantic/analyzed_type.py [hunk 4]
```diff
@@ -175,6 +248,16 @@ def json_schema(
         ref_template: str = DEFAULT_REF_TEMPLATE,
         schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
     ) -> dict[str, Any]:
+        """Generate a JSON schema for the model.
+
+        Args:
+            by_alias: Whether to use alias names (default: True).
+            ref_template: The format string used for generating $ref strings (default: DEFAULT_REF_TEMPLATE).
+            schema_generator: The generator class used for creating the schema (default: GenerateJsonSchema).
```
Comment (Kludex) [inline pydantic/analyzed_type.py:256]: There's also an inconsistency here: the type is not there, and the default value is written in parenthesis, while some are written after the period.

File: pydantic/color.py [hunk 5]
```diff
@@ -325,7 +406,17 @@ def parse_float_alpha(value: Union[None, str, float, int]) -> Optional[float]:
...
     """
-    Parse raw hue, saturation, lightness and alpha values and convert to RGBA.
+    Parse raw hue, saturation, lightness, and alpha values and convert to RGBA.
+
+    Args:
+        h: str, the hue value.
+        h_units: str, the unit for hue value.
+        sat: str, the saturation value.
+        light: str, the lightness value.
+        alpha: Optional[float], alpha value.
```
Comment (Kludex) [inline pydantic/color.py:416]: This is also a different format.

**Review by Kludex — COMMENTED** [review]
I think we should have a consistency on how to write the docstrings, even if we don't have a linter to enforce it for now.

File: pydantic/fields.py [hunk 6]
```diff
@@ -361,62 +361,49 @@ def Field(
...
-    :param regex: removed, use `Pattern` instead
-    :param **extra: deprecated, use `json_schema_extra` instead
+    Args:
+        default (Any): The default value is returned if the corresponding field value is not present in the input data
+            or if the data value is None. Defaults to `Undefined`.
+        default_factory (typing.Callable[[], Any] | None): A callable that returns the default value for the field.
+            Only used if default is not set. Defaults to `None`.
+        alias (str | None): The alias for the field. Defaults to `None`.
+        alias_priority (int | None): The priority score for the field if it is an alias for another field. Defaults to
+            `None`.
```
Comment (dmontagu) [inline pydantic/fields.py:371]: # TODO: Add documentation reference for non-str validation_alias variants
```suggestion
            `None`.
    validation_alias (str | list[str | int] | list[list[str | int]] | None): The alias(es) to use 
        to find the field value during validation. Defaults to `None`.
    serialization_alias (str | None): The alias to use as a key when serializing. Defaults to `None`.
```

**Review by Kludex — APPROVED** [review]

**Pull request merged**